test/e2e/upgrade_management_cluster_test.go

```diff
@@ -2,340 +2,137 @@ package e2e
-		clusterName         = "test1"
-		workloadClusterName = "test2"
-		clusterctlLogFolder string
-		kubernetesVersion   = e2eConfig.GetVariable(KubernetesVersion)
-	)
-	It("Should create a management cluster and then upgrade capi and capm3", func() {
-		/*---------------------------------------------*
-		| Create a target cluster with v1a4 clusterctl |
-		*----------------------------------------------*/
-		Logf("Starting v1a5 to v1b1 upgrade tests")
-
-		clusterctlBinaryURLTemplate := e2eConfig.GetVariable("INIT_WITH_BINARY")
-		clusterctlBinaryURLReplacer := strings.NewReplacer("{OS}", runtime.GOOS, "{ARCH}", runtime.GOARCH)
-		clusterctlBinaryURL := clusterctlBinaryURLReplacer.Replace(clusterctlBinaryURLTemplate)
-		clusterctlBinaryPath := "/tmp/clusterctltmp"
-
-		Logf("Downloading clusterctl binary from %s", clusterctlBinaryURL)
-		err := downloadFile(clusterctlBinaryPath, clusterctlBinaryURL)
-		Expect(err).ToNot(HaveOccurred(), "failed to download temporary file")
-		defer os.Remove(clusterctlBinaryPath) // clean up
-
-		err = os.Chmod(clusterctlBinaryPath, 0744)
-		Expect(err).ToNot(HaveOccurred(), "failed to chmod temporary file")
-
-		By("Creating a cluster")
-		Logf("Getting the cluster template yaml")
-		upgradeClusterTemplate := clusterctl.ConfigClusterWithBinary(ctx, clusterctlBinaryPath, clusterctl.ConfigClusterInput{
-			KubeconfigPath:       bootstrapClusterProxy.GetKubeconfigPath(),
-			ClusterctlConfigPath: clusterctlConfigPath,
-			// define template variables
-			Namespace:                namespace,
-			ClusterName:              clusterName,
-			KubernetesVersion:        kubernetesVersion,
-			Flavor:                   "ubuntu",
-			ControlPlaneMachineCount: pointer.Int64Ptr(1),
-			WorkerMachineCount:       pointer.Int64Ptr(1),
-			InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
-			// setup clusterctl logs folder
-			LogFolder: clusterctlLogFolder,
-		})
+})
 
-		Expect(upgradeClusterTemplate).ToNot(BeNil(), "Failed to get the cluster template")
+// Install the split-yaml tool to split our bmhosts_crs.yaml file.
+func installSplitYaml() {
+	cmd := exec.Command("bash", "-c", "kubectl krew update; kubectl krew install split-yaml")
+	output, err := cmd.CombinedOutput()
+	Logf("Download split-yaml:\n %v", string(output))
+	Expect(err).To(BeNil())
+}
 
-		Logf("Applying the cluster template yaml to the cluster")
+// Split the bmhosts_crs.yaml file to make it easy to apply limited number on BMH in each namespace.
+func splitBMHostFile() {
+	cmd := exec.Command("bash", "-c", "cat bmhosts_crs.yaml | kubectl split-yaml -t {{.name}}.yaml -p.")
+	cmd.Dir = workDir
+	output, err := cmd.CombinedOutput()
+	Logf("splitting bmhs:n %v", string(output))
+	Expect(err).To(BeNil())
+}
 
-		Expect(bootstrapClusterProxy.Apply(ctx, upgradeClusterTemplate)).To(Succeed())
+func getPreWaitForCluster(ctx context.Context) func(proxy framework.ClusterProxy, clusterNamespace string, clusterName string) {
```

Comment:
This needs a doc string and a better name. What is it doing?
Do you think it would be possible to define this as a variable instead before the `Describe`? I mean something like this:

```go
createBMHs := func(clusterProxy framework.ClusterProxy, clusterNamespace string) {...}
// ...
// Then use it like this
PreInit: createBMHs
```

Reply:
added doc I prefer to keep the generic name as it may do more than just creating bmh for example exporting some vars and to be consistent with other hooks

Reply:
Ok my point is not the outer function. It is clear that it just returns the inner function. My point is the inner function has no description for what it is doing and also no name so there is no way to understand it other than reading the code. It is not necessary to do it in this way. You can define the function separately and give it a name, then just pass it to the upgrade spec directly.
Like this:

```go
func myFancyPreWaitForClusterFunction(clusterProxy framework.ClusterProxy, clusterNamespace string, clusterName string) {
...
}

var _ = Describe("When testing cluster upgrade v1alpha5 > current [upgrade]", func() {
	capi_e2e.ClusterctlUpgradeSpec(ctx, func() capi_e2e.ClusterctlUpgradeSpecInput {
		return capi_e2e.ClusterctlUpgradeSpecInput{
			E2EConfig:                 e2eConfig,
			ClusterctlConfigPath:      clusterctlConfigPath,
			BootstrapClusterProxy:     bootstrapClusterProxy,
			ArtifactFolder:            artifactFolder,
			SkipCleanup:               true, // parameterize after merging https://github.com/kubernetes-sigs/cluster-api/pull/7373
			InitWithProvidersContract: "v1alpha4",
			InitWithBinary:            "https://github.com/kubernetes-sigs/cluster-api/releases/download/v0.4.7/clusterctl-{OS}-{ARCH}",
			PreInit:                   getPreInitFunc(ctx),
			PreWaitForCluster:         myFancyPreWaitForClusterFunction,
			PreUpgrade:                getPreUpgrade(ctx),
			MgmtFlavor:                osType,
			WorkloadFlavor:            "ubuntu",
		}
	})
})
```